test/test_autograd.py

```diff
@@ -1361,100 +1351,60 @@ def backward(self, grad_output):
-        y = torch.normal(x)
-        self.assertRaises(ValueError, lambda: y.reinforce(torch.randn(4, 5)))
-
-    def test_stochastic(self):
-        x = Variable(torch.rand(2, 10), requires_grad=True)
-        stddevs = Variable(torch.rand(2, 10) * 5, requires_grad=True)
-        y = (x * 2).clamp(0, 1)
-        y = y / y.sum(1, True).expand_as(y)
-        samples_multi = y.multinomial(5)
-        samples_multi_flat = y[0].multinomial(5)
-        samples_bernoulli = y.bernoulli()
-        samples_norm = torch.normal(y)
-        samples_norm_std = torch.normal(y, stddevs)
-        z = samples_multi * 2 + 4
-        z = z + samples_multi_flat.unsqueeze(0).expand_as(samples_multi)
-        z = torch.cat([z, z], 1)
-        z = z.double()
-        z = z + samples_bernoulli + samples_norm + samples_norm_std
-        last_sample = torch.normal(z, 4)
-        z = last_sample + 2
-        self.assertFalse(z.requires_grad)
+        self.assertEqual(logpmf.data[0], math.log(0.7 if s.data[0] == 1 else 0.3))
 
-        self.assertRaises(RuntimeError, lambda: z.backward(retain_graph=True))
-        samples_multi.reinforce(torch.randn(2, 5))
-        self.assertRaises(RuntimeError, lambda: z.backward(retain_graph=True))
-        samples_multi_flat.reinforce(torch.randn(5))
-        self.assertRaises(RuntimeError, lambda: z.backward(retain_graph=True))
-        samples_bernoulli.reinforce(torch.randn(2, 10))
-        self.assertRaises(RuntimeError, lambda: z.backward(retain_graph=True))
-        samples_norm.reinforce(torch.randn(2, 10))
-        self.assertRaises(RuntimeError, lambda: z.backward(retain_graph=True))
-        samples_norm_std.reinforce(torch.randn(2, 10))
-        # We don't have to specify rewards w.r.t. last_sample - it doesn't
-        # require gradient
-
-        last_sample.backward(retain_graph=True)
-        z.backward()
-
-        self.assertGreater(x.grad.data.abs().sum(), 0)
-
-    def test_stochastic_require_grad(self):
-        # This tests a DSD function sequence (D=deterministic, S=stochastic),
-        # where all functions require grad.
-        x = Variable(torch.randn(2, 10), requires_grad=True)
-        y = Variable(torch.randn(2, 10), requires_grad=True)
-        z = torch.normal(x + 2, 2)
-        o = z + y
-        z.reinforce(torch.randn(2, 10))
-        o.sum().backward()
-        self.assertEqual(y.grad.data, torch.ones(2, 10))
-        self.assertGreater(x.grad.data.abs().sum(), 0)
+        def apply_fn(p):
+            return torch.reinforce._bernoulli_logpmf(s, p)
+
+        gradcheck(apply_fn, (p,), raise_exception=True)
+
+    def test_reinforce_multinomial_1d(self):
+        p = Variable(torch.Tensor([0.1, 0.2, 0.3]), requires_grad=True)
+        s, logprobs = torch.reinforce.multinomial(p)
+        self.assertEqual(s.size(), logprobs.size())
+
+        def apply_fn(p):
+            return torch.reinforce._multinomial_logpmf(s, p)
+
+        gradcheck(apply_fn, (p,), raise_exception=True)
```

Comment:
why can't we test `torch.reinforce.multinomial` directly? Doesn't gradcheck skip outputs that don't require grad?

Reply:
The problem is the log-probabilities depend on the drawn sample. Consider bernoulli with p=0.3:

```
bernoulli(p) = (0, log(0.7))

bernoulli(p+eps) = (0, log(0.7-eps))
bernoulli(p+eps) = (1, log(0.3+eps))
```

If the subsequent calls to the random function return a different sample, then the gradchecks would fail.

Reply:
Makes sense, thank you :+1: